### PR TITLE
PNode sons: tidy and optimise

### PR DESCRIPTION
**Pull request by mbaulch**
Main points:
1. `ast.add` and `ast.addSons` have the same implementation. Replace calls of the former with the latter.
2. Pre-allocate `PNode` sons where possible. This reduces the compiler build time by approx 2%.

Nb: This patch reduces the size of the compiler binary too. (Around 8kb difference on my system)

### CODE REVIEW

**Comment by Araq** [comment]
Sorry, but ``addSon`` is a leftover from the Pascal days and ``add`` should be used consistently everywhere.

**Comment by mbaulch** [comment]
Done. Have renamed the other add procs, and those related to len/del/last for consistency.

File: compiler/ast.nim [hunk 1]
```diff
@@ -1232,11 +1227,9 @@ proc newSons*(father: PType, length: int) =
   else:
     setLen(father.sons, length)
 
-proc sonsLen*(n: PType): int = n.sons.len
 proc len*(n: PType): int = n.sons.len
-proc sonsLen*(n: PNode): int = n.sons.len
```
Comment (dom96) [inline compiler/ast.nim]: Is there a ``len`` defined for ``PNode``?

File: compiler/ast.nim [hunk 2]
```diff
@@ -1561,18 +1554,11 @@ proc isEmptyType*(t: PType): bool {.inline.} =
   ## 'void' and 'stmt' types are often equivalent to 'nil' these days:
   result = t == nil or t.kind in {tyVoid, tyStmt}
 
-proc makeStmtList*(n: PNode): PNode =
```
Comment (dom96) [inline compiler/ast.nim]: What's the rationale for removing this?

**Comment by dom96** [comment]
This is difficult to review because it's huge :(

**Comment by dom96** [comment]
Not sure it's a good idea to mix optimisations with so much renames.

**Comment by Araq** [comment]
This PR is way too large and has conflicts.

**Pull request closed without merging**